enforce unique map keys in CBOR decoding, ver. 9

JaredCorduan opened the pull request:
# Description

Starting in major version 9, we raise a CBOR decoding error if a map contains duplicate keys.

resolves #2965

# Checklist

- [x] Commit sequence broadly makes sense
- [x] Commits have useful messages
- [x] New tests are added if needed and existing tests are updated
- [x] Any changes are noted in the [changelog](https://github.com/input-output-hk/cardano-ledger/blob/master/CHANGELOG.md)
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (which can be run with `scripts/fourmolize.sh`
- [x] Self-reviewed the diff

lehins commented on libs/cardano-ledger-binary/src/Cardano/Ledger/Binary/Decoding/Decoder.hs:
```diff
@@ -831,6 +835,20 @@ decodeMapNoDuplicates decodeKey decodeValue =
+  (numValues, xs) <- decodeCollectionWithLen decodeMapLenOrIndef decodeInlinedPair
+  unless (numValues == (length . Set.toList . fromList . (map fst)) xs) $
+    fail "Map has duplicates"
+  pure $ Map.fromList xs
```
This is way too expensive:
* construct a list
* construct another list with keys
* construct a set from a list
* construct a map from a list

There is a much better way to do this.

Construct a Map and each time a key value pair is decoded check weather it is already in a Map or not.
*Disclaimer* - In PR programming, expect some problems, but the idea should be clear.

```suggestion
  decodeMapLenOrIndef >>= \case
    Just len -> loop (\(x -> pure (x - 1, x  <= 0)) len Map.empty
    Nothing -> loop (\x -> (,) x <$> decodeBreakOr) () Map.empty
  where
    loop condition prevStep !acc = do
      (nextStep, shouldStop) <- condition prevStep
      if shouldStop
        then pure acc
        else do
          (k, v) <- decodeInlinedPair
          when (k `Map.member` acc) $ fail "Duplicate key detected in the Map"
          loop condition nextStep (Map.insert k v acc)
```

JaredCorduan replied:
awesome, thank you!

lehins commented on libs/cardano-ledger-binary/test/Test/Cardano/Ledger/Binary/Vintage/Failure.hs:
```diff
@@ -42,6 +45,16 @@ genNegativeInteger :: Gen Integer
+  c <- Gen.int Range.constantBounded -- pick a value for duplicate key
+  Gen.shuffle ((a, c) : xs)
+  where
+    genPair = (,) <$> (Gen.int Range.constantBounded) <*> (Gen.int Range.constantBounded)
```
```suggestion
    genPair = (,) <$> Gen.int Range.constantBounded <*> Gen.int Range.constantBounded
```

lehins commented on libs/cardano-ledger-binary/test/Test/Cardano/Ledger/Binary/Vintage/Failure.hs:
```diff
@@ -80,6 +93,12 @@ prop_shouldFailNegativeNatural = property $ do
+prop_shouldFailMapWithDupKeys :: Property
+prop_shouldFailMapWithDupKeys = property $ do
+  xs <- forAll genDuplicateAssocList
+  let encodedMap = encodeMapLen (fromIntegral $ length xs) <> mconcat (toCBOR <$> xs)
```
I am not sure what's the name of function in Hedgehog, but something an equivalent to `elements` from QuickCheck can be used to switch between indefinite and exact length encoding:
```suggestion
  encodedMap <- 
    elements [ encodeMapLen (fromIntegral $ length xs) <> foldMap toCBOR xs
             , encodeMapLenIndef <> foldMap toCBOR xs <> encodeBreak ]
```

lehins reviewed (changes requested):
There is a more efficient way to implement stricter Map decoder, which I suggested in a comment.

Also, maybe you wanna try adding a quickcheck property test to hspec test suite? I think it would be a good exercise to get first hand experience with `hspec`.  This is just a suggestion, so feel free to ignore it.

Hedgehog definitely works, so the test that is already included in this PR will  work just fine, but since we mostly use QuckCheck, I thought it would be better if we stick to one tool for consistency. However, if you prefer Hedgehog there is nothing wrong with using both of the random property testing libraries.

JaredCorduan commented:
> Also, maybe you wanna try adding a quickcheck property test to hspec test suite? I think it would be a good exercise to get first hand experience with `hspec`. This is just a suggestion, so feel free to ignore it.

I only used Hedgehog since that was what was already being used in the `Failure` module. In general, I'd like to move completely over to QC.. I'll definitely poke around with using QuickCheck in hspec tomorrow, probably in a new test module, thanks for the motivation.

SebastienGllmt commented:
![](https://media0.giphy.com/media/5mBE2MiMVFITS/giphy.gif)

Commit:
enforce unique map keys in CBOR decoding, ver. 9

Starting in major version 9, we know raise a decoding error if a CBOR
map contains duplicate keys.

resolves #2965

JaredCorduan commented:
@lehins I redid the tests in QuickCheck. let me know if this is what you had in mind.

lehins reviewed (approved):
Beautiful!!! :slightly_smiling_face:

jmhrpr commented:
What was the behaviour when decoding a map with duplicate keys? Like, what value for the duplicate key was used?

JaredCorduan commented:
> What was the behaviour when decoding a map with duplicate keys? Like, what value for the duplicate key was used?

it was overwriting (so right biased).
